### DISCUSSION TITLE
Official docker images

### DISCUSSION THREAD

**Discussion by cproensa** [nicotine-plus/nicotine-plus#2312] [Ideas]
Is there any published docker image for nicotine+?
I find very convenient the usage of dockerized apps, for example for headless servers or NAS, or just for the ease of installation...

So far I was using [this docker image](https://github.com/realies/soulseek-docker) for soulseekQT, whis is very convenient:
minimal ubuntu LTS system, with openbox that only runs the maximized app, plus VNC server with web access.

For a long time i wanted to switch to Nicotine+, but not finding a good dockerized replacement has taken me back.
I would find very helpful if an official docker image, updated with the ongoing nicotine+ releases, can be offerd.

**Comment by slook** [comment]
> very convenient the usage of dockerized apps, for example for headless servers or NAS

It sounds like you might find `slskd` to be more suitable for this use case, as it does offer official dockers and is really intended for headless servers with a web UI and bunch a advanced client-agent features that will probably never be matched by Nicotine+.

See: https://github.com/slskd/slskd#quick-start

**Reply by EsEnZeT** [reply]
thanks for sharing, looks interesting

**Comment by cproensa** [comment]
> It sounds like you might find slskd to be more suitable for this use case

I have used that. In this scenario, is perfect for for the "uploads", but for the "downloads" it has a very lacking UI. And a lot less features.
Also, i'd rather have a client runing in a 24/7 server to leverage the wishlist search and download queues (and slskd does not have the autosearch feature)

I often see that SoulseekQT docker as the default recommendation for users with a NAS, becasue its the more used and tested option out there. This kind of user is very valuable: large storage to share, full time connected...
It's a pity that there is not a nice N+ image, stable and supported.

I managed to modify that image to have nicotine+ running. Apart from features or personal preferencies, it has much better performance, less RAM and CPU usage overall...
If i have the time, i may try to learn the proper workflows for building images, and create and publish a clean one.
Any help is appreciated

**Reply by mathiascode** [reply]
I don't use Docker, but I could publish official images as long as the following points are considered:
- Support both Docker and Podman (and other available alternatives?)
- Either use Wayland or Broadway, avoid X11. https://docs.gtk.org/gtk4/broadway.html
- Support GTK 4
- Avoid unnecessary cruft and dependencies, make image as minimal and easy to maintain as possible
- Make the image as straightforward as possible to set up and use

**Reply by cproensa** [reply]
I am no expert in any way... but always is a good time to learn.
My intention if time allows it, is to setup a basic image, which should follow your guidelines. If the result is good, it can be a start to make it available totally 3rd party, unofficially supported, etc.
With time, and if there is good traction, a more serious integration with the project can be devised.

Expect me to ask some technical questions along the way for N+ best practices and your recommendations, as needed.

**Reply by slook** [reply]
> it can be a start to make it available totally 3rd party, unofficially supported, etc.

My worry here is that an unofficial release would be more likely to become abandoned at some point, or a dud version would get stuck out in the wild for years.

> With time, and if there is good traction, a more serious integration with the project can be devised.

It would be preferred if yourself and the developers who contribute to the future maintenance of a Docker image could pull actions runners or some other mechanism from the outset in order to automate releases out of the official repo alongside the other installation methods.

**Reply by mathiascode** [reply]
Here's some initial work in the `packaging/container/` folder. I haven't tested it yet, but let me know if it works: https://github.com/mathiascode/nicotine-plus/tree/docker-image

If it builds and starts, it should be accessible in a web browser on the 8090 port.

**Reply by EsEnZeT** [reply]
I was able to access container after adding some changes:
- https://github.com/mathiascode/nicotine-plus/pull/9

```
➜  container git:(docker-image) ✗ pwd
/tmp/nicotine-plus/packaging/container
➜  container git:(docker-image) ✗ docker compose build; docker compose up
```

The problem is:
<img width="1324" alt="image" src="https://user-images.githubusercontent.com/3882467/212207353-3dd820cf-a565-4f9a-b5ee-e05c98aeb4b9.png">

I saw something similar, that some window(s) are outside of my view on other unofficial containers long time ago, but I'm not sure how to deal with that.

I didn't test further than that.

**Reply by slook** [reply]
A similar centered-window position bug is known on certain systems possibly related https://github.com/nicotine-plus/nicotine-plus/discussions/1881#discussioncomment-2798560 it looks like we cannot know the docker viewport size during init.

You can Escape the broken Setup Assistant dialog then login after using the Preferences dialog to set username and pass etc?

@cproensa Please can you test the development branch https://github.com/mathiascode/nicotine-plus/tree/docker-image to verify if it is working the same on your docker?

**Reply by mathiascode** [reply]
The other dialogs are centered, so this seems to be https://github.com/nicotine-plus/nicotine-plus/discussions/1881#discussioncomment-2798560 again... The dialog seems to be centered at (0, 0) this time though.

**Reply by slook** [reply]
@EsEnZeT The dialog centering issue is fixed in master and the development branch now, you could test that connecting and everything else works?

**Reply by EsEnZeT** [reply]
@slook sure, I will be happy to! Will try to do it tomorrow. Sorry for slow response but unfortunately was sick.

//EDIT: I checked now (it was quick anyways) and I don't see any changes from what I saw before unfortunately.


```
➜  container git:(docker-image) git --no-pager log --decorate=short --pretty=oneline -n1
dfdc10cfb1d6a65106db4884fb03b2ea38a6dd52 (HEAD -> docker-image, origin/docker-image) Merge branch 'nicotine-plus:master' into docker-image
```

`docker compose build --no-cache && docker compose up`

<img width="1318" alt="image" src="https://user-images.githubusercontent.com/3882467/214462291-d31aaf62-e9ae-4d3e-95b0-2e4f8657c533.png">

Looks the same on multiple browsers

**Reply by slook** [reply]
Can you capture the keyboard and Escape the dialog, then can you interact with the main window and the Preferences dialog normally, or is that affected also?

What happens when you unmaximize the main window, does it restore into the centre, or is it affected aswell?

**Reply by EsEnZeT** [reply]
> Can you capture the keyboard and Escape the dialog, then can you interact with the main window and the Preferences dialog normally, or is that affected also?

When I click ESC to get rid of the incorrectly centered 'welcome window' I can interact just fine with everything else I guess.
When I click ESC then (x) to quit (restart) app it's actually restarted with 'welcome window' being centered correctly.

> What happens when you unmaximize the main window, does it restore into the centre, or is it affected aswell?

Unmaximize - window to the left a bit; I can move it, change position etc, no issues at all
Unmaximize -> Maximize restores just fine

**Reply by slook** [reply]
> When I click ESC then (x) to quit (restart) app it's actually restarted with 'welcome window' being centered correctly.

By that point, the default main window position and size is set correctly in the config. It is up to the environment's "window manager" to decide how dialog should be centered, either on parent window or centered on screen. Clearly, it's choosing to calculate on the parent window size (which is 0 x 0 pixels before it's drawn).

It seems like we are still calling the welcome dialog too early in the startup sequence, i.e. before the main window is fully realized.

**Reply by mathiascode** [reply]
> It seems like we are still calling the welcome dialog too early in the startup sequence, i.e. before the main window is fully realized.

Not our fault. Depending on if broadway is already running and visible in the browser, the dialog is either properly centered, centered at (0, 0), or the whole window is misplaced like this:
![](https://user-images.githubusercontent.com/8754153/214622830-7163b5e7-bf19-4f89-b6f0-890e96127785.png)

**Reply by slook** [reply]
> Not our fault

No, it's not our fault, clearly this is an upstream bug.

> Depending on if broadway

Is there any way to detect this type of environment at runtime, such that evasive action can be taken, such as a maximized dialog, or detect if the position is outside the viewport during init, to at least make sure the widgets inside it are usable.

**Comment by toineenzo** [comment]
Regarding my closed issue #2548 , i'll repost this here:
For me Nicotine+ is the best client to use for SLSK. I'm running it right now in a docker container on my NAS to always have it available. I also would like to connect to it from outside my network. I could secure it via Cloudflare, but it would be much easier if there was an option in the N+ settings to password protect the app/ui. I could use [slskd ](https://github.com/slskd/slskd)but it doesn't have an easy (settings) UI. 
As @mathiascode said: _Perhaps HTTP authentication support should be added to the Docker image instead?_
That would be nice to include if the official docker images (ever) get released. :)

**Comment by cproensa** [comment]
> For me Nicotine+ is the best client to use for SLSK. I'm running it right now in a docker container on my NAS to always have it available.

Thats my use case. 
So if you need authentication, IMO the place should be at any of the higher layers: vnc, http proxy, vpn, etc
Personally i went for simplicity and used a cloudflare tunnel to accesss this dockerized vnc+nicotine, and no need to expose any ports on my public ip.

_And as i feared, i've had still little time to do a decent repo for the image i'm using..._ 

Regarding the proof of concept in https://github.com/mathiascode/nicotine-plus/tree/docker-image
My issue with the Broadway through html presentation, is that web browsers dont work well with copy & paste.
For my intensive use of nicotine+, i usually connect directly through a pure vnc client, which works perfect for bidirectional copy and paste.

**Reply by toineenzo** [reply]
I'm curious how you got the vnc/ websocket connection working with cloudflare. I kept getting a blank screen upon accessing my domain via a websocket page and with authentication via Cloudflare Access.

**Reply by toineenzo** [reply]
![image](https://github.com/nicotine-plus/nicotine-plus/assets/4921397/e4cc0e6f-7244-4f33-beb1-139eecea226b)
This is the error in the console btw. I set SSL to Full Strict btw and enabled the websocket option in cloudflare settings.

**Reply by cproensa** [reply]
I'm sorry i can't help you much on that. When i need to connect with Cloudflare access i have a tunnel directly to a "webtop" image serving a desktop in the browser, and then use that as a jump to any other app inside my network,

**Comment by cproensa** [comment]
So i built my [image ](https://hub.docker.com/r/cproensa/nicotine-docker) with nicotine and novnc, using ubuntu with the .deb package. And i'm having this issue with (most of) the icons not showing:
![n1](https://github.com/nicotine-plus/nicotine-plus/assets/14123811/c6ae085b-3310-4ad0-9c51-66d579c32678)
![n2](https://github.com/nicotine-plus/nicotine-plus/assets/14123811/5fddee9a-5e7b-41c2-a7bc-a965112b9922)

I added all the packages that were listed as dependecies in the .deb package.
Is there a specific package that enables those icons?
if so... is there a missing dependency in the nicotine deb?
This happens with the bare ubuntu base image. Of course, when installing the .deb in a full ubuntu installation everything works right..
@slook @mathiascode any suggestion?

**Reply by cproensa** [reply]
Found it:
`librsvg2-common`

still:  Shuold be that a dependecy in the -deb package?

**Reply by slook** [reply]
Sorry I have no concrete suggestions for you @cproensa . You are the one whose knowledge is most superior on this topic.

I'd imagine a separate packaging script will be required for the build process that is specific to the planned docker image release. However, if you feel that the main project dependencies are incorrect then feel free to submit a PR so it can be reviewed/investigated as appropriate.

Merging https://github.com/cproensa/nicotine-docker/commit/f412aa247e7ddbe3da5f5e163d24916011dd1401 with the main nicotine-plus repo is unlikely to do any harm, so push whatever solution you desire that makes life easy.

> So i built my [image](https://hub.docker.com/r/cproensa/nicotine-docker)

Nice job on adapting the documentation by the way. Your repo looks tidy.

> https://github.com/cproensa/nicotine-docker#how-to-build 

Can the process be automated? Do you have any experience with GitHub Actions?

**Reply by cproensa** [reply]
>  You are the one whose knowledge is most superior on this topic.

Absolutely not. This is my first attempt at building images, and still, its an adaptation of someone else's project.

> if you feel that the main project dependencies are incorrect then feel free to submit a PR so it can be reviewed/investigated as appropriate.

not an expert in the packaging topics, but i can take it as another thing to learn and investigate

> Merging https://github.com/cproensa/nicotine-docker/commit/f412aa247e7ddbe3da5f5e163d24916011dd1401 with the main nicotine-plus repo is unlikely to do any harm

Before that, I'd rather do more testing to reproduce and discard some kind of mistake on my side

> Can the process be automated? Do you have any experience with GitHub Actions?

At this moment i don't intend to automate or go into CI processes. No experience, either. Maybe in the future...
The most i will do is build stable and nightly images locally, for my use. Very convenient to quickly test the new unreleased feautres. Having the images already, its trivial to push them to the public hub. But not a full CI.

Anyway, the proper way to integrate is what Mat did in his repo. I tested his container (which uses broadway) and is perfectly good but i'm missing the raw vnc. As I explained before, modern browsers restricts the clipboard interaction, which is a pain...

**Reply by mathiascode** [reply]
> Anyway, the proper way to integrate is what Mat did in his repo. I tested his container (which uses broadway) and is perfectly good but i'm missing the raw vnc. As I explained before, modern browsers restricts the clipboard interaction, which is a pain... 

As soon as I have some more time, I could look into finishing my implementation and merging it with this repo, as well as the clipboard stuff.

Regarding the missing dependencies, I'm not sure if other GTK applications explicitly mention them. I'll have to check what other packages do.

**Comment by slook** [comment]
Has anybody in here tested Nicotine+ 3.3.0.dev6 in Docker recently? We're planning on bumping to Release Candidate very soon.

**Comment by cproensa** [comment]
> Has anybody in here tested Nicotine+ 3.3.0.dev6 in Docker recently? 

I am runing this:
https://hub.docker.com/r/cproensa/nicotine-docker/
with regular updates to the latest testing deb package

> We're planning on bumping to Release Candidate very soon.

During the latest weeks i have seen a variation of errors:
- crash with an error message
- freezes
- instant app shutdown without any error or warnings

But i assume its due to the unstability of running the latest dev branch, and some of those errors come and go as i update.
I think, once its into RC status, i should open issues for any of those it they happen again.

**Reply by mathiascode** [reply]
Please open issues, otherwise things will never get fixed.

**Comment by mathiascode** [comment]
For the record, an official Docker image won't happen with the release of Nicotine+ 3.3.0, but it's something we can consider later, if third-party images work well.

**Comment by toineenzo** [comment]
Please take a look at https://hub.docker.com/r/sirjmann92/nicotineplus-proper and https://github.com/sirjmann92/nicotineplus-proper which is a Docker image that uses the Broadway backend of GTK for the display server. This makes it much much faster and stable then noVNC. It also has a branch tag that uses GTK4 that works pretty good.

**Comment by slook** [comment]
Nicotine+ 3.3.7rc1 release candidate will be released soon. Please verify compatibility of containers with latest master and let us or the image maintainers know if there's any issues or whatnots.

**Reply by toineenzo** [reply]
@sirjmann92

**Comment by fletchto99** [comment]
I've got a container here based off of the linuxserver selkies images: https://github.com/fletchto99/nicotine-plus-docker - Open to feedback but it should be working well from the latest stable nicotine plus release. Published to dockerhub and GHCR
